Add in java column to row conversion [skip ci]

revans2 opened the pull request:
This should be ready to go for reviews and checking in. This is also my first cuda code so I would love feedback on it.  I have working code for fixed width on the plugin and it massively speeds up transitions to/from the GPU when there are multiple columns involved.

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

hyperbolic2346 commented on java/src/main/native/src/row_conversion.cu:
```diff
@@ -0,0 +1,365 @@
+        rmm::mr::device_memory_resource* mr
+        ) {
+    int64_t total_allocation = size_per_row * num_rows;
+    if (total_allocation > std::numeric_limits<int>::max()) {
```
why not make this `CUDF_EXPECTS(total_allocation < std::numeric_limits<int>::max(), "Table is too large to fit!");`

hyperbolic2346 commented on java/src/main/native/src/row_conversion.cu:
```diff
@@ -0,0 +1,365 @@
+    // If we don't have enough shared memory there is no point in haveing more threads per block that will just sit idle
+    max_block_size = max_block_size > x_possible_block_size ? x_possible_block_size : max_block_size;
+    int block_size = (max_block_size / 32) * 32;
+    if (block_size == 0) {
```
`CUDF_EXPECTS(block_size != 0, "Row size is too large to fit in shared memory");`

revans2 commented:
@hyperbolic2346 I have addressed your comments.  I also added in the conversion back from rows to columns.  I have not tested the performance on this one yet so I am going to keep this as work in progress until I can do a better job with that.

razajafri commented:
Overall Java looks OK to me. 

Could we say that we are essentially serializing the table, as Row_0| validity_0 |P| Row_1| validity_1|P| ... | Row_N|validity_N|P

revans2 commented:
> Overall Java looks OK to me.
> 
> Could we say that we are essentially serializing the table, as Row_0| validity_0 |P| Row_1| validity_1|P| ... | Row_N|validity_N|P

I'll try to clarify it a bit.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/HostColumnVectorCore.java:
```diff
@@ -121,7 +121,7 @@ public long getNullCount() {
    */
-  List<HostColumnVectorCore> getNestedChildren() {
+  public List<HostColumnVectorCore> getNestedChildren() {
     return children;
```
Question- why do we need this change? I thought we did not want to expose this as getting to a child was available thru ColumnViewAccess APIs.

revans2 replied:
You are right.  I did it while prototyping changes to the plugin.  I'll fix it.

kuhushukla commented on CHANGELOG.md:
```diff
@@ -21,6 +21,7 @@
 - PR #6555 Adapt JNI build to libcudf composition of multiple libraries
 - PR #6564 Load JNI library dependencies with a thread pool
+- PR #6578 Add in java column to row conversion
 
```
Can we fix the PR title to the text in the change log?

kuhushukla commented on java/src/main/java/ai/rapids/cudf/Table.java:
```diff
@@ -1619,6 +1623,105 @@ public Table gather(ColumnVector gatherMap, boolean checkBounds) {
+   *
+   * The format of each row is similar in layout to a C struct where each column will have padding
+   * in front of it to align it properly. Each row has padding inserted at the end so the next row
+   * is aligned to a 64-bit boundary. This is so the first column will always start at the beginning
```
nit : `so that the` ?

kuhushukla reviewed (commented):
Minor stuff else lgtm!

revans2 commented on java/src/main/java/ai/rapids/cudf/HostMemoryBuffer.java:
```diff
@@ -639,14 +639,27 @@ public final HostMemoryBuffer sliceWithCopy(long offset, long len) {
    * WARNING: Debug only method to print a passed in buffer
    */
   public void printBuffer() {
+    printBuffer(5);
```
This is there to match what the old API did.  I didn't want to just remove it.

kuhushukla reviewed (approved):
Looks good. I reviewed the have piece only though.

mythrocks commented on java/src/main/native/src/row_conversion.cu:
```diff
@@ -0,0 +1,636 @@
+        rmm::mr::device_memory_resource* mr) {
+    std::unique_ptr<rmm::device_uvector<T>> ret(new rmm::device_uvector<T>(
+                input.size(), stream, mr));
+    CUDA_TRY(cudaMemcpyAsync(ret->data(),
```
My understanding of asynchronous CUDA `memcpy()` is iffy. From [the manual](https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-async), I understand that this copy from non-pinned host memory to device-memory? Does any synchronization need to be done before the result of the copy is first accessed? (Or is it so fast that it doesn't really matter?)

revans2 replied:
No synchronization is needed if the copy and the kernel that reads the data both happen on the same stream.  CUDA guarantees FIFO order within a stream.  The time we would have to worry is when we are copying data to the host, because the host does not explicitly wait or if we will use a different stream to read the data.

mythrocks replied:
I see, thanks. Here's a bug that reminded me of this conversation:
https://github.com/rapidsai/cudf/issues/6700

mythrocks commented on java/src/main/native/src/row_conversion.cu:
```diff
@@ -0,0 +1,636 @@
+    if (num_blocks < 1) {
+        num_blocks = 1;
+    } else if (num_blocks > 10240) {
+        num_blocks = 10240;
```
This function is informative.

Why is there a cap (`10240`) for the number of blocks in the grid?

revans2 replied:
I picked an arbitrary magic number.  I should probably try to remove it because it does not matter.  When I first read about blocks it said that CUDA had a limit of 65,535 blocks in the x dimension, and I thought that hitting a high number like that probably added a lot of overhead, but it turns out that we require compute capability 6.0 or above which actually has a limit of `2^31 - 1` for the x dimension in blocks, so I should play around with it a bit more.  That is essentially the same as the number of rows that we can support in a single column.  I'll play around with it and see if I run into any issues if I let it go very large.

mythrocks commented on java/src/main/native/src/row_conversion.cu:
```diff
@@ -0,0 +1,636 @@
+ * @param [out] column_size the size in bytes of the data for each columns in the row.
+ * @return the size in bytes each row needs.
+ */
+static inline int64_t compute_fixed_width_layout(
```
Nitpick: I have missed the reason that this returns an `int64_t`, when `at_offset` is 32bit, as is the return type for `align_offset()`.

revans2 replied:
There is no reason, I should fix it.

mythrocks commented on java/src/main/native/src/row_conversion.cu:
```diff
@@ -0,0 +1,636 @@
+    // need to change if we split a word of validity data between blocks.
+    int y_block_size = (num_columns + 3) / 4;
+    if (y_block_size > 32) {
+      y_block_size = 32;
```
I might have misunderstood, but Line#204 indicates:
```
    // We do not support copying a subset of the columns in a row yet, so we don't
    // currently support a row that is wider than shared memory
```
But we're also capping the block's Y-dimension at 32. Should we add `CUDF_EXPECTS()` to bail out if the number of columns in the table exceeds 32?

revans2 replied:
Sorry I should update the comments here. Each Y dimension ideally handles about 4 columns.  This is what I found from tests on a V100 to be ideal.  No idea if it holds for other architectures too.  So we first try to get a y dimension where each thread is going to read/write about 4 columns.  If that fails then we let the number of columns increase above 4.  So we can handle an arbitrary number of columns.  What we cannot handle is a row that is larger than fits in shared memory.

For performance reasons we go through shared memory as we transform between rows/columns so all of the main memory accesses are coalesced. We also want to have at most 48KB of shared memory.  That limit varies some what between different compute capabilities but 48KB is the smallest amount supported by what we support so I went with that.  Also to take full advantage of coalescing reads/writes I want to make sure that the X dimension is always a multiple of 32.  This means that each row can be at most 48/32 or 1.5KB.

That is enough for 1365 byte columns or 722 short columns or 372 int columns or 189 long columns or depending on padding/etc some odd combination of all of them.  There is a check further down for this.

razajafri commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -165,6 +165,7 @@ synchronized int getRefCount() {
    */
   public ColumnVector copyToDevice() {
     if (rows == 0) {
+      // TODO this does not work for nested types!!!
```
Should we throw here if it's a nested type?

revans2 replied:
The exception is thrown by the constructor that is being called.  I found this because of a bug in my code that caused an empty column to be copied to the device, which should never happen.

razajafri reviewed (changes requested):
Please check my comment above. 

https://github.com/rapidsai/cudf/pull/6578/files#r512436976

revans2 commented:
@razajafri please take another look

razajafri reviewed (approved):
LGTM

revans2 commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -3689,4 +3689,39 @@ void testStructColumnFilterStrings() {
+        assert rows.length == 1;
+        ColumnVector cv = rows[0];
+        assert cv.getRowCount() == t.getRowCount();
+//        try (HostColumnVector hcv = cv.copyToHost()) {
+//          hcv.getChildColumnViewAccess(0).getDataBuffer().printBuffer(8);
+//        }
+
```
Yes it helps with debugging.
